dds/idl/topic_keys.cpp

```diff
@@ -0,0 +1,500 @@
+        element_count_ *= dimension;
+      }
+    }
+    AST_Type* type_node = array_node->base_type();
+    AST_Type* unaliased_type_node = type_node->unaliased_type();
+    for (; pos_ < element_count_; ++pos_) {
+      child_ = new Iterator(unaliased_type_node, this);
+      Iterator& child = *child_;
+      if (child == Iterator()) {
+        delete child_;
+        child_ = 0;
+        throw Error(array_node, "array type is marked as key, but its base type "
+          "does not contain any keys.");
+      }
+      current_value_ = *child;
+      return *this;
+    }
+
+  // If we are a union, use self if we have a key
+  } else if (root_type_ == UnionType) {
+    if (pos_ == 0) { // Only Allow One Iteration
+      pos_ = 1;
+      AST_Union* union_node = dynamic_cast<AST_Union*>(root_);
+      if (be_global->has_key(union_node)) {
+        current_value_ = root_;
+        return *this;
+      } else {
+        throw Error(union_node, "union type is marked as key, "
+          "but it's discriminator isn't");
+      }
+    }
+
+  // If we are a primitive type, use self
+  } else if (root_type_ == PrimitiveType) {
+    if (pos_ == 0) { // Only Allow One Iteration
+      pos_ = 1;
+      current_value_ = root_;
+      return *this;
+    }
+  }
+
+  // Nothing left to do, set this to null
+  *this = end_value();
+
+  return *this;
+}
+
+TopicKeys::Iterator
+TopicKeys::Iterator::operator++(int)
+{
+  Iterator prev(*this);
+  ++(*this);
+  return prev;
+}
+
+AST_Decl*
+TopicKeys::Iterator::operator*() const
+{
+  return current_value_;
+}
+
+bool
+TopicKeys::Iterator::operator==(const TopicKeys::Iterator& other) const
+{
+  return
+    parent_ == other.parent_ &&
+    root_ == other.root_ &&
+    root_type_ == other.root_type_ &&
+    pos_ == other.pos_ &&
+    current_value_ == other.current_value_ &&
+    level_ == other.level_ &&
+    recursive_ == other.recursive_ &&
+    dimensions_ == other.dimensions_ &&
+    (
+      (child_ && other.child_) ? *child_ == *other.child_ : child_ == other.child_
+    );
+}
+
+bool
+TopicKeys::Iterator::operator!=(const TopicKeys::Iterator& other) const
+{
+  return !(*this == other);
+}
+
+std::string
+TopicKeys::Iterator::path()
+{
+  std::stringstream ss;
+  path_i(ss);
+  return ss.str();
+}
+
+void
+TopicKeys::Iterator::path_i(std::stringstream& ss)
+{
+  if (root_type_ == StructureType) {;
+    AST_Structure* struct_root = dynamic_cast<AST_Structure*>(root_);
+    AST_Field** field_ptrptr;
+    struct_root->field(field_ptrptr, child_ ? pos_ : pos_ - 1);
+    AST_Field* field = *field_ptrptr;
+    ss << (level_ ? "." : "") << field->local_name()->get_string();
+  } else if (root_type_ == UnionType) {
+    // Nothing
+  } else if (root_type_ == ArrayType) {
+    // Figure out what the vector version of the scalar pos_ is
+    std::vector<size_t>::reverse_iterator di, dfinished = dimensions_.rend();
+    size_t acc = pos_;
+    size_t div = 1;
+    std::list<size_t> results;
```

Comment:
Does this need a `std::list`?  Looks like it always adds to one side and then iterates last-in first-out.  The equivalent with vector is push_back and reverse iterators.  Not that we're so performance-sensitive inside the compiler...

Reply:
Yes I can change that. I guess I automatically thought list when I had to prepend elements.